From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Sleigh-InSPECtor <sleighinspector@outlook.com>
Date: Tue, 28 May 2024 17:09:24 +0930
Subject: [PATCH] 6590: AArch32: ldrexd addr src register aliased as first dest
 register causing second word load at the value of first load

AArch32: fixed ldrexd

* fixed addr src register aliased as first dest register of a double word load causing second word load at the value of first load
---
 Ghidra/Processors/ARM/data/languages/ARMinstructions.sinc | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/Ghidra/Processors/ARM/data/languages/ARMinstructions.sinc b/Ghidra/Processors/ARM/data/languages/ARMinstructions.sinc
index b32f0b4137..e69f6df455 100644
--- a/Ghidra/Processors/ARM/data/languages/ARMinstructions.sinc
+++ b/Ghidra/Processors/ARM/data/languages/ARMinstructions.sinc
@@ -3654,9 +3654,10 @@ ArmPCRelImmed12: reloff		is U23=0 & immed & rotate
 
 :ldrexd^COND Rd,Rd2,[Rn]  is $(AMODE) & ARMcond=1 & COND & c2027=0x1b & Rn & Rd & Rd2 & c0011=0xf9f
 {
-	build COND;
-  	Rd = *(Rn);
-  	Rd2 = *(Rn + 4);
+  build COND;
+  local addr:4 = Rn;
+  Rd = *(addr);
+  Rd2 = *(addr + 4);
 }
 
 :ldrexh^COND Rd,[Rn]  is $(AMODE) & ARMcond=1 & COND & c2027=0x1f & Rn & Rd & c0011=0xf9f
-- 
2.45.1

